### PR TITLE
Adding authentication and role-based access control

### PR DESCRIPTION
**Pull request by misa1313**
Addressing issue #13

### CODE REVIEW

File: oauth.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+Handles token creation and verification. 
+"""
+from datetime import datetime, timedelta
+from jose import jwt, JWTError
+from fastapi import Depends, HTTPException
+from fastapi.security import OAuth2PasswordBearer
+from settings import SECRET_KEY
+from schemas import TokenData, UserRole
+
+oauth2_scheme = OAuth2PasswordBearer(tokenUrl='login')
```
Comment (zklevsha) [inline oauth.py]: tokenUrl should match yout login enpoin ('/api/v1/login')
Without it swagger ui cant authenticate you

File: tasks_crud.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+crud.py
+This module handles all the functions called by the app.py module, 
+as well as managing the functionality of the DB operations.
+"""
+import sqlalchemy as sa
+from fastapi import HTTPException
+from sqlalchemy.ext.asyncio import AsyncSession, AsyncEngine
+from sqlalchemy import text
+from models import Todo
+from users_crud import handle_errors
```
Comment (zklevsha) [inline tasks_crud.py]: handle_errors are used in two modules tasks_crud.py and users_crud.py
I think it`ll be better to move this function to separate helpers.py module

Also why not create folder/package crud and move tasks_crud and users_crud.py here
i.e.
```
├── crud
│   ├── __init__.py
│   ├── tasks.py
│   └── users.py

```

File: settings.py [hunk 3]
```diff
@@ -3,8 +3,35 @@
 Module for getting all the required
 connection details of the databases.
 """
+import random
+import string
+import bcrypt
 from dotenv import dotenv_values
 
+def generate_random_string(length):
```
Comment (zklevsha) [inline settings.py]: move those functions to helpers.py or crypto.py module
no need to keep it in settings

File: oauth.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
+"""
```
Comment (zklevsha) [inline oauth.py:1]: There is deprecation warning I thing it`ll needs to be addressed
```
oauth.py:21: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
    expiration = datetime.utcnow() + timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
```

File: tests/test_tasks.py [hunk 5]
```diff
@@ -1,9 +1,10 @@
 """
-test_app.py
-The main module containing all tests for the FastAPI application.
+test_tasks.py
+The module containing all task-related tests for the FastAPI application.
 """
 from fastapi.testclient import TestClient
-from conftest import app, get_db, get_engine #pylint: disable=W0611
+from conftest import app, get_db, get_engine, generate_creds #pylint: disable=W0611
+from test_users import get_new_token
```
Comment (zklevsha) [inline tests/test_tasks.py]: maybe it`ll be better to move get_new_token to tests/helpers.py

File: tests/test_tasks.py [hunk 6]
```diff
@@ -1,9 +1,10 @@
 """
-test_app.py
-The main module containing all tests for the FastAPI application.
+test_tasks.py
```
Comment (zklevsha) [inline tests/test_tasks.py:2]: Maybe we also should test do admin related tasks 
```
Admins can modify or delete any tasks.
Only admins can create new admins.
```

File: tests/test_users.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    Function to generate a new user for the tests.
+    """
+    user, email = generate_creds()
+    user_data = {"username": user, "email": email, "password": "test"}
+    response = await test_client.post(f"{BASE_URL}/register", json=user_data)
+    parsed_response = UserOutput(**response.json())
+    return parsed_response
+
+
+async def log_user(test_client) -> None:
```
Comment (zklevsha) [inline tests/test_users.py]: it`ll betere to rename function to login
log looks like logging related

**Review by zklevsha — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -65,128 +81,89 @@ Run the FastAPI application using Uvicorn:
...
+| `GET`   | `/api/v1/tasks/{task_id}`  | Get Task by Id     | Yes                     |
+| `PUT`   | `/api/v1/tasks/{task_id}`  | Update Todo        | Yes                     |
+| `DELETE`| `/api/v1/tasks/{task_id}`  | Delete Todo        | Yes (Owner or Admin)    |
+| `PUT`   | `/api/v1/tasks/{task_id}/finish` | Mark Completed | Yes                   |
 
-Expected response:
+#### Users
 
-```json
-`{"message":"This is another test route."}`
```
Comment (zklevsha) [inline README.md:92]: Remove `
this symbols cause red marks on code example

File: crud/users.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+                detail='An internal server error occurred. Please try again later.'
+            ) from error
+        finally:
+            await db.close()
+
+    return wrapper
+
+
+@handle_errors
+async def create_new_user(user: UserCreate, db: AsyncSession):
```
Comment (zklevsha) [inline crud/users.py]: Currently, any user can register as an admin. We should allow users to register only as regular users. If needed, an admin can update a user's role and promote them to an admin.

File: schemas.py [hunk 10]
```diff
@@ -31,8 +36,90 @@ class TodoData(BaseModel):
...
+
+
+class UserCreate(BaseModel):
+    """
+    Model for the user creation endpoint.
+    """
+    username: str
+    email: EmailStr
+    password: Union[str, bytes]
+    creation_date: int = Field(default_factory=lambda: int(datetime.now().timestamp()))
```
Comment (zklevsha) [inline schemas.py]: User should not  set creation date. It should be set on backend itself
Same could aslo be done for Todos creation date

**Review by zklevsha — CHANGES_REQUESTED** [review]

File: tests/test_users.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
+"""
```
Comment (zklevsha) [inline tests/test_users.py:1]: Add tests for /register endpoint
Dont forget to check that only admin can create new admin

File: tests/test_users.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+    # Testing with the admin user.
+    headers = {
+        "Authorization": f"Bearer {ADMIN_TOKEN}"
+    }
+    response = await test_client.get(f"{BASE_URL}/", headers=headers)
+    assert response.status_code == 200
+
+
+async def test_delete_user(test_client) -> None:
```
Comment (zklevsha) [inline tests/test_users.py:122]: check that admin can delete other users

**Review by zklevsha — CHANGES_REQUESTED** [review]

File: crud/tasks.py [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+crud.py
+This module handles all the functions called by the app.py module, 
+as well as managing the functionality of the DB operations.
+"""
+import sqlalchemy as sa
+from fastapi import HTTPException
+from sqlalchemy.ext.asyncio import AsyncSession, AsyncEngine
+from sqlalchemy import text
+from models import Todo
+from crud.users import handle_errors
```
Comment (zklevsha) [inline crud/tasks.py]: move hande_errors to crud.helpers

**Review by zklevsha — CHANGES_REQUESTED** [review]

File: README.md (hunk 8 above)
Comment (misa1313) [reply README.md:92]: Hi! That seems to refer to an older version of the README file. The new one doesn't include that text.

**Pull request merged**